homeassistant/components/apcupsd/config_flow.py

```diff
@@ -0,0 +1,87 @@
+"""Config flow for APCUPSd integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.config_entries import CONN_CLASS_LOCAL_POLL, ConfigFlow
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import selector
+import homeassistant.helpers.config_validation as cv
+
+from . import DOMAIN, APCUPSdData
+
+_LOGGER = logging.getLogger(__name__)
+
+_PORT_SELECTOR = vol.All(
+    selector.NumberSelector(
+        selector.NumberSelectorConfig(
+            min=1, max=65535, mode=selector.NumberSelectorMode.BOX
+        ),
+    ),
+    vol.Coerce(int),
+)
+
+
+class ConfigFlowHandler(ConfigFlow, domain=DOMAIN):
+    """APCUPSd integration config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        schema = vol.Schema(
+            {
+                vol.Required(CONF_HOST, default="localhost"): cv.string,
+                vol.Required(CONF_PORT, default=3551): _PORT_SELECTOR,
+            }
+        )
+
+        if user_input is None:
+            return self.async_show_form(step_id="user", data_schema=schema)
+
+        # Test the connection to the hostand get the current status for serial number.
+        data_service = APCUPSdData(user_input[CONF_HOST], user_input[CONF_PORT])
+        try:
+            await self.hass.async_add_executor_job(data_service.update)
+        except OSError:
+            errors = {"base": "cannot_connect"}
+            return self.async_show_form(
+                step_id="user", data_schema=schema, errors=errors
+            )
+
+        if len(data_service.status) == 0:
+            return self.async_abort(reason="no_status")
+
+        # We _try_ to use the serial number of the UPS as the unique id since
+        # this field is not guaranteed to exist on all APC UPS models.
```

Comment:
Maybe also guard against more than one config entry with the same host and port, after checking for unique_id abort?

https://github.com/home-assistant/core/blob/cf5a11a1e7c7ef49a439753f3bb708d10b2c22b9/homeassistant/config_entries.py#L1281-L1285

Reply:
Wasn't aware of this, added the guard! 

It is impossible to have two apcupsd integrations with same host and port (while having different `unique_id`s), so I feel it is better to put the guard right after `user_input` is not `None`. 

(If a system has multiple APC UPSes, [there must be multiple instances of apcupsd listening on different addresses](https://wiki.debian.org/apcupsd#Set_up_the_apcupsd_configuration_files))

Let me know if you think putting it after unique_id abort is still a good idea.